### PR TITLE
Refactor unix domain socket store config

### PR DESCRIPTION
**Pull request by fzakaria**
# Motivation

Following what is outlined in #10766 refactor the uds-remote-store such that the member variables (state) don't live in the store itself but in the config object.

Additionally, the config object includes a new necessary constructor that takes a scheme & authority.

Minor:
* code formatting
* cleanup of getting default path
* added some comments

# Context
Please see  #10766
This adds part of the work for the unix domain socket store.

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: src/libstore/uds-remote-store.cc [hunk 1]
```diff
@@ -27,37 +26,25 @@ std::string UDSRemoteStoreConfig::doc()
...
-
 std::string UDSRemoteStore::getUri()
 {
-    if (path) {
-        return std::string("unix://") + *path;
-    } else {
-        // unix:// with no path also works. Change what we return?
-        return "daemon";
-    }
+    return std::format("{}://{}", scheme, getPathOrDefault());
```
Comment (Ericson2314) [inline src/libstore/uds-remote-store.cc:47]: We don't have this on macOS yet, I'm afraid. better just use `+`?

Comment (Ericson2314) [reply src/libstore/uds-remote-store.cc:47]: I am not sure whether we want to stop returning `"daemon"` when the path is not the default, either.

File: src/libstore/store-api.cc [hunk 2]
```diff
@@ -1289,7 +1289,12 @@ ref<Store> openStore(StoreReference && storeURI)
             if (access(stateDir.c_str(), R_OK | W_OK) == 0)
                 return std::make_shared<LocalStore>(params);
             else if (pathExists(settings.nixDaemonSocketFile))
-                return std::make_shared<UDSRemoteStore>(params);
+                // TODO(fzakaria): A bit gross that we now pass empty string
+                // but this is knowing that empty string will later default to the same
+                // nixDaemonSocketFile. Why don't we just wire it all through?
+                // I believe there are cases where it will live reload so we want to
+                // continue to account for that.
+                return std::make_shared<UDSRemoteStore>(UDSRemoteStoreConfig::scheme, "", params);
```
Comment (Ericson2314) [inline src/libstore/store-api.cc:1297]:
```suggestion
                return std::make_shared<UDSRemoteStore>(params);
```
oh sorry, you can keep the old constructor around, precisely for this use case. You can make the old constructor wrap the new one, i.e. `UDSRemoteStore(UDSRemoteStoreConfig::scheme, "", params)` *delegating constructor* call. And then `UDSRemoteStoreConfig::scheme` can go back to being private.

**Review by Ericson2314 — CHANGES_REQUESTED** [review]
It would be good to have a unit test like https://github.com/NixOS/nix/pull/11108, very close otherwise! :)

File: src/libstore/uds-remote-store.hh [hunk 3]
```diff
@@ -7,42 +7,67 @@
...
+    // outlined in https://github.com/NixOS/nix/issues/10766
+    using LocalFSStoreConfig::LocalFSStoreConfig;
+    using RemoteStoreConfig::RemoteStoreConfig;
+
+    UDSRemoteStoreConfig(std::string_view scheme, std::string_view authority, const Params & params)
         : StoreConfig(params)
         , LocalFSStoreConfig(params)
         , RemoteStoreConfig(params)
     {
+        assert(scheme == UNIX_SCHEME && "Scheme must be 'unix'");
```
[the remark this answers is not in the document]
Comment (Ericson2314) [reply src/libstore/uds-remote-store.hh]: I think you didn't do this one yet?

File: src/libstore/uds-remote-store.hh [hunk 4]
```diff
@@ -9,16 +9,37 @@ namespace nix {
 
 struct UDSRemoteStoreConfig : virtual LocalFSStoreConfig, virtual RemoteStoreConfig
 {
-    UDSRemoteStoreConfig(const Params & params)
+
+    static constexpr char const * scheme = "unix";
```
Comment (Ericson2314) [inline src/libstore/uds-remote-store.hh:13]: Can be `protected` now

**Comment by fzakaria** [comment]
Failing MacOS builds but not sure why -- appreciate any pointer.

File: src/libstore/uds-remote-store.hh [hunk 5]
```diff
@@ -9,16 +9,43 @@ namespace nix {
 
 struct UDSRemoteStoreConfig : virtual LocalFSStoreConfig, virtual RemoteStoreConfig
 {
-    UDSRemoteStoreConfig(const Params & params)
+
+    protected:
+        static constexpr char const * scheme = "unix";
+    
+    public:
```
Comment (Ericson2314) [inline src/libstore/uds-remote-store.hh:16]:
```suggestion
protected:
    static constexpr char const * scheme = "unix";
    
public:
```
it's a bit weird looking, but that's how this is generally formatted (pre and post the formatter).

**Review by Ericson2314 — APPROVED** [review]

**Pull request merged**